### ISSUE TITLE
Improve the memory footprint of a large number of mostly idle connections

### ISSUE THREAD

**Issue by larsfillmore** [rustls/rustls#794]
Our use case is a server that supports a large number of client connections that are usually idle or send/receive small messages (think of a chat server).

The C10k memory usage is about 300M, while switching to pure TCP reduces it to about 100M (BTW this corresponds well to https://jbp.io/2019/07/02/rustls-vs-openssl-memory-usage.html).

We're already using `Session::set_buffer_limit()`. But there are still internal buffers that are kept allocated even when there is no data to send/receive.

One possible solution might be to enable buffer reuse between connections. Something like `ServerConfig::buffer_provider:  Option<Arc<dyn BufferProvider>>`:
```rust
trait BufferProvider {
    fn acquire(&self) -> Buffer;
    fn release(&self, Buffer); 
}
```

The session would acquire a buffer once it sees activity on the underlying stream and release the buffer afterwards. The user may implement `BufferProvider` as a buffer pool.

That's rather a feature request than a design proposal as it doesn't touch related topics like reducing double-buffering (https://github.com/ctz/rustls/issues/716#issuecomment-851496486)

**Comment by djc** [comment]
Short term, we could provide a method that deallocates buffers that are no longer needed after the handshake, only after the handshake is complete. I'd be interested in your assessment of the impact of something like #796. Warning: that's based on the main branch, which contains a substantial number of API changes. On the plus side, it's very possible that the main branch already improves memory usage for you.

For the longer term, I think the ideas discusses in #716 and #668 will likely yield an API that gives you enough control over memory allocation. If you'd be interested in helping fund that work, I'd be happy to chat about that.

**Comment by briansmith** [comment]
> Short term, we could provide a method that deallocates buffers that are no longer needed after the handshake, only after the handshake is complete.

Why not just deallocate those buffers immediately when the handshake completes, automatically? It seems wrong to keep them around after the handshake.

**Comment by larsfillmore** [comment]
> I'd be interested in your assessment of the impact of something like #796. Warning: that's based on the main branch, which contains a substantial number of API changes.

I didn't try applying #796 yet, sorry. It will be interesting to compare, but I need to first back-port it to 0.19.1 and implement an auto-minimize to make it useable with my Cargo.toml/[replace].

Nevertheless, the main contributor to the memory consumption seems to be [`MessageDeframer::buf: Box<[u8; OpaqueMessage::MAX_WIRE_SIZE]>`](https://github.com/rustls/rustls/blob/5e9ccd4b5b1154ac4b55316dfad178ba09957069/rustls/src/msgs/deframer.rs#L21), where `OpaqueMessage::MAX_WIRE_SIZE` is about 18K.

I experimented with allocating `MessageDeframer::buf` on demand and destroying it once it is not used (in 0.19.1): https://gist.github.com/larsfillmore/41ce8aace8f454a5dbb04f2e9110c18c. This worked as expected and reduced my app memory consumption from 3x to 1.5x compared to raw TCP.

**Comment by djc** [comment]
Did you also measure the effect on performance (throughput/latency)?

**Comment by larsfillmore** [comment]
> Did you also measure the effect on performance (throughput/latency)?

No, I did not. I don't think that re-creating a buffer for each received message would be acceptable from the performance POW in the general case. Hence the idea to let the user implement a buffer pool if needed.

> My changes don't affect the deframer, since (I think?) that is still used after the handshake.

Yes, it is used after the handshake.

> Also, my patch does affect state that is visible in the public API, so doing all of that minimization cannot be triggered automatically (but some of it can be).

I see. My code uses `rustls` indirectly via `tokio-rustls` and therefore it will take some effort to test non-automatic minimization.

Most probably we'll need both: (1) the minimization after the handshake completion and (2) not keeping the `MessageDeframer::buf` while there is nothing to read/reconstruct.

**Comment by djc** [comment]
> No, I did not. I don't think that re-creating a buffer for each received message would be acceptable from the performance POW in the general case. Hence the idea to let the user implement a buffer pool if needed.

Yes, but getting caller allocation designed right is pretty involved, so it will take some more time. I'd probably not be inclined to add API surface for doing reallocations like this in the meantime.

> I see. My code uses rustls indirectly via tokio-rustls and therefore it will take some effort to test non-automatic minimization.

You can still call `TlsStream::get_mut()` and call `minimize()` on that, right?

**Comment by larsfillmore** [comment]
> You can still call TlsStream::get_mut() and call minimize() on that, right?

Indeed. Thank you!

So, I back-ported your patch to 0.19.1:
```rust
impl ServerSession {
    pub fn minimize(&mut self) -> bool {
        if self.is_handshaking() {
            log::trace!("Not minimizing");
            return false;
        }

        log::trace!("Minimizing");

        self.imp.common.handshake_joiner = crate::msgs::hsjoiner::HandshakeJoiner::new();
        self.imp.alpn_protocol = None;
        self.imp.client_cert_chain = None;
        self.imp.sni = None;
        self.imp.received_resumption_data = None;
        self.imp.resumption_data = Vec::new();
        true
    }
}
```

Then I added the calls of `ServerSession::minimize()` to my app and checked that the log contained "TRACE rustls::server: Minimizing".

The result: there is no noticeable difference in the memory usage (on macOS + jemalloc). I find this surprising.

**Comment by djc** [comment]
In some basic testing, replacing the `HandshakeJoiner` (in `process_handshake_messages()`, after `is_handshaking()` is `false`) caused some issues with our TLS compliance tests, so not sure whether that change is actually correct, either.

**Comment by gustav3d** [comment]
> jemalloc)
Jemalloc got configurable strategy for returning memory to the OS.

**Comment by briansmith** [comment]
https://github.com/rustls/rustls/issues/850#issuecomment-999798205 contains some suggestions that, if implemented, should allow much less RAM to be used, as all certificate-related RAM could be discarded after verification.

**Comment by horazont** [comment]
Hmm, that sounds neat. Though in my benchmarking I found the never-released 18 kiB buffer more impactful than anything else allocated by rustls.

**Comment by larsfillmore** [comment]
> Though in my benchmarking I found the never-released 18 kiB buffer more impactful than anything else allocated by rustls.

True. But once the impact of this 18K buffer is reduced, optimising other sources of RAM usage may become important.

**Comment by SSebo** [comment]
@larsfillmore Hi,  I encounter the same problem. Do you find a work around solution?

**Comment by larsfillmore** [comment]
> @larsfillmore Hi, I encounter the same problem. Do you find a work around solution?

Not a good one.

If you're willing to trade CPU for MEM you may try doing https://gist.github.com/larsfillmore/41ce8aace8f454a5dbb04f2e9110c18c (https://github.com/rustls/rustls/issues/794#issuecomment-896213768).

(This issue is not pressing for us yet, so I didn't give much attention to it)

**Comment by SSebo** [comment]
@djc Can we use `bytes` crate to dynamically allocate buffer, and shrink to fit after some period?  not fixed [u8] for `[MessageDeframer::buf: Box<[u8; OpaqueMessage::MAX_WIRE_SIZE]>](https://github.com/rustls/rustls/blob/5e9ccd4b5b1154ac4b55316dfad178ba09957069/rustls/src/msgs/deframer.rs#L21)`

**Comment by djc** [comment]
There's no need to use the bytes crate, we could probably just use a `Vec`. @SSebo would you like to submit a PR?

**Comment by larsfillmore** [comment]
@djc I don't think using a `Vec` will help. `MessageDeframer` will still need to either keep an allocated `Vec` between messages or will need to alloc/zeroinit/dealloc it at most of `MessageDeframer::read()`.

What do you think of the following approach?:
* Keep a thread-local `Box<[u8; OpaqueMessage::MAX_WIRE_SIZE]>`.
* Read into it in [`MessageDeframer::read()`](https://github.com/rustls/rustls/blob/0a5c9776345f29b79fc9fed2024b131fd70abfb3/rustls/src/msgs/deframer.rs#L58).  
  If nothing left in it after the [`try_deframe_one()` loop](https://github.com/rustls/rustls/blob/0a5c9776345f29b79fc9fed2024b131fd70abfb3/rustls/src/msgs/deframer.rs#L68), do nothing else (presumably, the most frequent case). Otherwise copy the unparsed bytes into a freshly-allocated `MessageDeframer::partial_msg: Vec<u8>` and reuse them in a successive `MessageDeframer::read()`.

If you're Ok with this I'll submit a PR.

**Comment by SSebo** [comment]
@djc I try to use `Vec` to reduce MessageDeframer mem usage, but the memory consumption and performance are worse than  @larsfillmore 's implementation.

**Comment by djc** [comment]
#1171 should improve on this situation.

**Comment by larsfillmore** [comment]
> #1171 should improve on this situation.

I did some testing using the `main` branch (which includes #1171). There is a massive improvement compared to `0.20.8`. My app's memory consumption has dropped by more than 50%.

Thank you!

Now that the buffer size is dynamic, we may try and add buffer auto-shrinking as mentioned in https://github.com/rustls/rustls/pull/1171#discussion_r1070984963. There seems to be a very simple algorithm for this, but I need to do more perf testing.

**Comment by djc** [comment]
Great to hear!

**Comment by gustav3d** [comment]
Great work !!

**Comment by larsfillmore** [comment]
> Now that the buffer size is dynamic, we may try and add buffer auto-shrinking as mentioned in [#1171 (comment)](https://github.com/rustls/rustls/pull/1171#discussion_r1070984963). There seems to be a very simple algorithm for this, but I need to do more perf testing.

#1180 seems to do the thing.

**Comment by bdaehlie** [comment]
It seems this issue has been addressed so closing it. If there are other improvements people would like to see please few new more specific issues. Thanks!
